r/persistent_volume - add support for CSI

DrFaust92 opened the pull request:
### Description

<!--- Please leave a helpful description of the pull request here. --->

### Acceptance tests
- [x] Have you added an acceptance test for the functionality being added?
- [x] Have you run the acceptance tests on this branch? (If so, please include the test log in a gist)

```
--- PASS: TestAccKubernetesPersistentVolume_csi_basic (3.65s)
```

### References

<!---
Are there any other GitHub issues (open or closed) or pull requests that should be linked here? Vendor blog posts or documentation?
--->
Relates #296

benosman commented on kubernetes/schema_volume_source.go:
```diff
@@ -218,6 +218,41 @@ func commonVolumeSources() map[string]*schema.Schema {
+					},
+					"volume_handle": {
+						Type:        schema.TypeString,
+						Description: "A map that specifies static properties of a volume. More info: https://kubernetes.io/docs/concepts/storage/volumes/#csi",
```
This description is incorrect - should refer to a string like:
`A string value that uniquely identifies the volume. `

DrFaust92 replied:
Fixed

benosman commented:
I'm just an end user, but I tried out this pull request and with my initial testing, so far it's working as it should for me.

DrFaust92 commented:
Thanks for the input @benosman

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/terraform-provider-kubernetes?pullRequest=817)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/terraform-provider-kubernetes?pullRequest=817)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/terraform-provider-kubernetes?pullRequest=817).</sub>

jrhouston commented on kubernetes/structure_persistent_volume_spec.go:
```diff
@@ -1518,6 +1559,28 @@ func patchPersistentVolumeSource(pathPrefix, prefix string, d *schema.ResourceDa
+		newV, newOk := newIn.([]interface{})
+
+		if newOk && len(newV) > 0 {
+			if oldOk && len(oldV) > 0 {
```
I think we could clean the duplication here by setting 

```go
path := pathPrefix + "/csi"
value := expandCSIPersistentDiskVolumeSource(newV)
```

before the conditional, and then just supply them when constructing AddOperation/ReplaceOperation

DrFaust92 replied:
should i just do this for all other as well? maybe extract to a more general function.
id addressed the changes you mentioned, i can open another PR for a more general refactor

jrhouston commented on kubernetes/schema_volume_source.go:
```diff
@@ -218,6 +218,41 @@ func commonVolumeSources() map[string]*schema.Schema {
+			Description: "Represents a CSI Volume. More info: http://kubernetes.io/docs/user-guide/volumes#csi",
+			Optional:    true,
+			MaxItems:    1,
+			Elem: &schema.Resource{
```
I see in the [API documentation](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#csipersistentvolumesource-v1-core) for this resource there are several additional attributes for referencing secrets - is there a reason we haven't added them here?

DrFaust92 replied:
they are a bit tricky but i added them.
while the secret_ref state that `namespace` is optional. the api call fails if it not set. i extracted the `secret_ref` block to a common func (maybe better to extract to provider wide later)
i kept the argument optional as stated in the docs (to also make the block common). its a non breaking change as the namespace is also computed.

DrFaust92 commented:
addressed comments + some "minor" changes.

tests output:
```
--- PASS: TestAccKubernetesPersistentVolume_csi_basic (5.32s)
--- PASS: TestAccKubernetesPersistentVolume_csi_importBasic (2.89s)
--- PASS: TestAccKubernetesPersistentVolume_csi_secrets (2.69s)
```

DrFaust92 commented:
@jrhouston, polite nudge
